### PR TITLE
Expose dirent pointer and added readDirStreamWith

### PR DESCRIPTION
**Pull request by mmhat**
This PR adds two new functions `readDirStreamWith` and `readDirStreamWithPtr`:
 * `readDirStreamWith` is a version of `readDirStreamMaybe` that takes a callback that is used to obtain the result from the pointer to the directory entry. This directory pointer is wrapped in a newtype called `DirEnt`. The function is intended to be used in cases where we know about the features of the installed libc: For example, glibc provides the information about the type of the directory entry directly in the dirent struct. Therefore it can be used directly without calling further functions and additional calls to stat(2).
Since those extensions to dirent are not covered by POSIX but widely used, we provide only the possibility to make use of those and do not add this functionality to `unix`.
 * `readDirStreamWithPtr` takes a pre-allocated pointer in addition to the other arguments of `readDirStreamWith`. This pointer is used to store the pointer of the dirent struct (if there is any) which can safe some allocations when you want to read a lot of directory entries as this pre-allocated pointer can be shared between the calls to the function.

In addition to that a small CPP bug was fixed which prevented compilation on my system. Also, some tests for the readDirStream functions were added.

### CODE REVIEW

**Comment by vdukhovni** [comment]
Why not conditionally provide a new high-level Linux/glibc-specific API that returns a directory entry *and* the  file type.  The conditional test here would be the availability of the `DT_REG` and `DT_UNKNOWN` entry types.  The module would then also expose pattern synonyms for the expected types.

I think this would be cleaner/safer than exposing the proposed API.  As for avoiding allocation of a dirent structure, with all the attendant syscall overhead, the memory GC cost should be fairly modest.

So I am not entirely enthused by this.  Which is not a hard no, but significant scepticism.

**Comment by hasufell** [comment]
Am more in favor of https://hackage.haskell.org/package/hpath-posix-0.13.3/docs/System-Posix-RawFilePath-Directory-Traversals.html#v:readDirEnt

**Comment by mmhat** [comment]
@vdukhovni 

> Why not conditionally provide a new high-level Linux/glibc-specific API that returns a directory entry _and_ the file type. The conditional test here would be the availability of the `DT_REG` and `DT_UNKNOWN` entry types. The module would then also expose pattern synonyms for the expected types.

It was my understanding that the unix package is about POSIX-compliance and additional features should be implemented in upstream libraries like hpath-path. This patch aims to make it easier for upstream libraries to do that as they don't have to copy/replicate the work done here.
If the conditional high-level API you proposed would be accepted as a contribution I'd be more than happy to implement those.

> 
> I think this would be cleaner/safer than exposing the proposed API. As for avoiding allocation of a dirent structure, with all the attendant syscall overhead, the memory GC cost should be fairly modest.

Ok, that's fair. I consider `readDirStreamWithPtr` as optional and I am willing to remove this function if requested

**Comment by mmhat** [comment]
@hasufell 

> Am more in favor of https://hackage.haskell.org/package/hpath-posix-0.13.3/docs/System-Posix-RawFilePath-Directory-Traversals.html#v:readDirEnt

I am aware of `hpath-posix` (Great library!) and the purpose of this patch is precisely to make libraries like this one easier to implement. For example, I noticed that the function linked by you is essentially a copy of the `readDirStream` + `d_type`. My hope is that this contribution will reduce the need for code duplication like this.

File: tests/ReadDirStream.hs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+dir :: FilePath
+dir = "dir"
+
+emptyDirStream :: IO ()
+emptyDirStream = do
+  cleanup
+  createDirectory dir ownerReadMode
+  dir_p <- openDirStream dir
+  _ <- readDirStreamMaybe dir_p  -- Just "."
+  _ <- readDirStreamMaybe dir_p  -- Just ".."
```
[the remark this answers is not in the document]
Comment (mmhat) [reply tests/ReadDirStream.hs]: Oh, no, that shouldn't be this way. I implemented your suggestions.

**Comment by hasufell** [comment]
I'm not opposed to this. Can we vote on this @hs-viktor @vdukhovni @Bodigrim 

:+1: from me

**Comment by vdukhovni** [comment]
> I'm not opposed to this. Can we vote on this @hs-viktor @vdukhovni @Bodigrim
> 
> 👍 from me

Which `this` are you in favour of? As I said, I'd rather not expose the proposed API and instead support the higher-level interface.

**Comment by hasufell** [comment]
> > I'm not opposed to this. Can we vote on this @hs-viktor @vdukhovni @Bodigrim
> > 
> > 👍 from me
> 
> Which `this` are you in favour of? As I said, I'd rather not expose the proposed API and instead support the higher-level interface.

The changes in this PR. I'm fine with these low-level interfaces. Alternatively, we can also put them into **Internal** module and expose it to signal "you're on your own".

**Comment by hasufell** [comment]
If no other maintainer chimes in, then I'll merge this in 2 weeks.

**Comment by Bodigrim** [comment]
@hs-viktor @vdukhovni just a gentle reminder.

**Comment by vdukhovni** [comment]
I still mostly stand by my original comment, and perhaps expose this only as an Internal API with no stability guarantees. Maintainers of dependent packages would have to be willing to adapt to future changes...

Why not just support the higher-level (system-specific) API?

**Comment by hasufell** [comment]
I'm fine with exposing this as Internal.

**Comment by hasufell** [comment]
https://github.com/haskell/ghcup-hs/issues/415#issuecomment-1344184716

One more reason to merge this and add tests.

**Comment by Bodigrim** [comment]
This is way out of my area of expertise, but I'm inclined to agree with Viktor.

> It was my understanding that the unix package is about POSIX-compliance and additional features should be implemented in upstream libraries like hpath-path. This patch aims to make it easier for upstream libraries to do that as they don't have to copy/replicate the work done here.

Practically `unix` is a package for everything which is not Windows. E. g., WASM is not a POSIX-compliant platform, and almost every OS has its own set of quirks violating POSIX. 

> If the conditional high-level API you proposed would be accepted as a contribution I'd be more than happy to implement those.

I'd accept such contribution.

**Comment by Bodigrim** [comment]
@mmhat are you interested to work on a high-level API here?

**Comment by hasufell** [comment]
I don't think it's appropriate for this PR to provide the high-level API. We're going to burn out the contributor this way. We should decide whether we merge this PR mostly as-is and expose it as internal API.

This has a :+1: from me.

**Comment by Bodigrim** [comment]
> We're going to burn out the contributor this way.

That's a very sensible concern, but @mmhat originally said that "If the conditional high-level API you proposed would be accepted as a contribution I'd be more than happy to implement those". It seems a win-win to me.

**Comment by mmhat** [comment]
> > We're going to burn out the contributor this way.
> 
> That's a very sensible concern, but @mmhat originally said that "If the conditional high-level API you proposed would be accepted as a contribution I'd be more than happy to implement those". It seems a win-win to me.

@Bodigrim Luckly I have a bit more time to work on this until the end of the year so I could do that.

**Comment by mmhat** [comment]
> Why not conditionally provide a new high-level Linux/glibc-specific API that returns a directory entry _and_ the file type. The conditional test here would be the availability of the `DT_REG` and `DT_UNKNOWN` entry types. The module would then also expose pattern synonyms for the expected types.
> 
> I think this would be cleaner/safer than exposing the proposed API. As for avoiding allocation of a dirent structure, with all the attendant syscall overhead, the memory GC cost should be fairly modest.
> 
> So I am not entirely enthused by this. Which is not a hard no, but significant scepticism.

@Bodigrim @hasufell I added some high-level API as outlined in the comment by @vdukhovni. I also moved `readDirStreamWith` and `readDirStreamPtr` to the internal API. Can you have a quick look if this first draft matches your expectations?

**Comment by hasufell** [comment]
I just stumbled over this:

* https://github.com/haskell/ghcup-hs/issues/766
* https://stackoverflow.com/a/29628894

**Comment by vdukhovni** [comment]
Another reason to not expose the **low-level** API.

**Comment by hasufell** [comment]
In ghcup I've written a portable version of `readDirStream`: https://github.com/haskell/ghcup-hs/pull/767/commits/53e324bfee681290c87d467d54142a9c15bba387#diff-2d9015010bcf8511a8be4190d7203f50fea2ce7f5b04d3f2d44bea2b93914ce1R116

This is similar to what other projects have been doing (falling back to 'stat' on 'DT_UNKNOWN'): https://github.com/ggreer/the_silver_searcher/pull/37

**Comment by mmhat** [comment]
> Another reason to not expose the **low-level** API.

I kindly disagree: IMHO, when it comes to directory tree traversal, the functions we provide in this package are _already_ low-level. The user needs to understand concepts like a directory stream and we throw a bunch of low-level technical terms at them in the Haddocks (like `dirent`, `d_name`, ...). I think it is reasonable to assume that they are familiar with all that stuff in general and that exposing `DirType` does not add another level of "low-leveliness" that should be hidden.

To resolve this issue with unsupported filesystems I see the following options:

1. We decide that exposing `DirType` comes indeed with too much limitations and we do not want to support it in this library. Hence this contribution here is rejected.
2. We expose the `readDirStreamWithType` functions and `DirType` as part of an _internal_ API. That would probably mean to create modules like `System.Posix.Directory.Internals.ByteString` and move those functions there.
3. We expose the `readDirStreamWithType` functions and `DirType` as part of an _public_ API as it is currently done in this PR.

If we go with (2) or (3) I'd like to add some documentation pointing out what the limitations of `DirType` are. In addition to that I could add a portable version of `readDirStreamWithType` like @hasufell did for ghcup-hs or change it to `readDirStreamWithType :: Bool -> ...` where the `Bool` indicates whether to use `stat` on `DT_UKNOWN` or not. Another option is to return a `Maybe DirType` where `Nothing` means `DT_UNKNOWN`.

Personally, I'd prefer option (3) and I'd rather add the `Bool` argument than introduce another (portable) version of `readDirStreamWithType`.

**Comment by hasufell** [comment]
> where the `Bool` indicates whether to use `stat` on `DT_UKNOWN`

When would someone pass `False` here? I believe we can expose the non-portable version in a module that's clearly marked non-portable. I think otherwise the unix package really strives to be... well, portable.

No need for a Bool option imo.

**Comment by mmhat** [comment]
> > where the `Bool` indicates whether to use `stat` on `DT_UKNOWN`
> 
> When would someone pass `False` here? I believe we can expose the non-portable version in a module that's clearly marked non-portable. I think otherwise the unix package really strives to be... well, portable.
> 
> No need for a Bool option imo.

Good point; I always thought of the `readDirStream` functions as wrappers around `readdir`, hence being unopinionated about fallback behavior and leave it up to the user to decide what to do if a filesystem doesn't support certain functionality. So far for the theory, I admit I don't have a use case in mind where somebody would pass `False`.

There is however a corner case I am not sure how to deal with: Suppose the underlying platform does not properly support `d_type` and `DT_UNKNOWN`, `DT_REG` and friends are not defined in `dirent.h`. Then the corresponding constants in this package will be all be `-1` due to `FP_CHECK_CONSTS` in the configure.ac and a fallback implementation like the one linked in https://github.com/haskell/unix/pull/251#issuecomment-1416136882 won't save us.

**Comment by hasufell** [comment]
> There is however a corner case I am not sure how to deal with in a fallback implementation: Suppose the underlying platform does not properly support d_type and some of DT_UNKNOWN, DT_REG, ... are not defined in dirent.h. Then the corresponding constants in this package will be all be -1 due to FP_CHECK_CONSTS in the configure.ac and a fallback implementation like the one linked in #251 (comment) won't save us as the constructed DirType value is meaningless.

We should be able to check for DT_UNKNOWN **or** -1 in the stat-fallback, no?

**Comment by mmhat** [comment]
@hasufell I guess I didn't explain my concerns very well; What I meant was: Suppose neither `DT_REG` nor `DT_DIR` are defined, hence `pattern RegularFileType = DirType (-1)` and `pattern DirectoryType = DirType (-1)`. We encounter a regular file in the directory stream and since no known `DirType` pattern matches we decide to `stat` it. Since it is a regular file we return `RegularFileType` to the user. But when the user checks if the entry is a directory, i.e. `case dtype of { DirectoryType -> ...; RegularFileType -> ...}`, this check will be successful as `RegularFileType == DirectoryType`.

The (well, obvious) solution to that is to assign different (negative) values to `CONST_DT_*` if those are missing on the underlying platform. I implemented that and the desired fallback behavior and added a new 'DirStreamWithPath` type that stores the filepath that corresponds to the directory stream. Please let me know what you think of it.

**Comment by mmhat** [comment]
BTW: The CI is currently failing with "Error: All install methods for ghc 8.10.7 failed". I didn't look into it yet, but I suspect https://github.com/haskell/unix/pull/270 will resolve this issue?

**Comment by Bodigrim** [comment]
@mmhat Seems intermittent, CI is functional again. Please fix a build error in the GHC 8.2 build.

**Comment by Bodigrim** [comment]
@hasufell @vdukhovni I'll leave this for you to review, I'm absolutely illiterate in this area.

**Review by vdukhovni — COMMENTED** [review]
I haven't fully digested this PR as yet. It looks mostly reasonable, but I have one area of concern:

The use of `DirStreamWithPath` is IMHO a bit too explicit. I'd like this representation to be more opaque. The caller should not have to know or be able to rely on the fact that the directory stream knows its (original at open time) location.

The use of this type is only as a fallback for lack of file type info in the directory entry, and the saved path is used to perform `stat(2)` operations on the relative names of each entry. There are two reasons to avoid exposing this fact.

1. Some systems (sufficiently modern Linux) provide `statx(2)` which can perform a stat relative to a directory file handle, which resolves various potential TOCTOU issues when traversing a volatile directory tree, and should also be more performant.  On such systems we don't want to use the saved path, and should ultimately use `statx(2)` instead.

2. For now I'd be inclined to **NOT** automatically simulate availability of file types in the directory entry. Instead, just return unknown/unavailable, and let the caller decide whether and how to perform the additional stat. They may not want to do the stat at all in many cases (when it is not *free*).

My preferred solution is "2", but otherwise, can the representation fo the *enriched* DirStream be more opaque, the caller should not know how it is augment (if at all) to perform file type deduction.

**Comment by hasufell** [comment]
> For now I'd be inclined to NOT automatically simulate availability of file types in the directory entry. Instead, just return unknown/unavailable, and let the caller decide whether and how to perform the additional stat.

That's not a high-level API anymore, which is what you were asking for if I remember correctly.

I don't like APIs that are in-between.

Either expose the syscalls with minimal wrapper and all the caveats or provide a high-level API that is as cross platform as it can be.

**Comment by vdukhovni** [comment]
> > For now I'd be inclined to NOT automatically simulate availability of file types in the directory entry. Instead, just return unknown/unavailable, and let the caller decide whether and how to perform the additional stat.
> 
> That's not a high-level API anymore, which is what you were asking for if I remember correctly.
> 
> I don't like APIs that are in-between.
> 
> Either expose the syscalls with minimal wrapper and all the caveats or provide a high-level API that is as cross platform as it can be.

I don't agree that returning *unavailable* for the file type, when not returned by the directory traversal is a failure to offer a high-level API. Sure, it doesn't completely hide the system behaviour, but to me that's a feature.

If there's a desire to unconditionally get the file type, that can be another function layered on top, and the user can provide either a path or file handle relative to which to interpret the directory entries.  Then there's no need for the DirStreamWithPath type, and the resolution of the entries to file types, sizes, ... is an orthogonal matter, often best handled via `statx(2)` and friends, or if the intent is to open the files anyway, then `openat` and `fstat`, ...

In other words, the extra logic to try to stat the files when the data is not directly available is both potentially wasteful and likely not as robust as what the application should actually do.

**Comment by mmhat** [comment]
> My preferred solution is "2"

I'd also go with option 2 here, but I also see how a `stat` fallback can be valuable in many cases. How about the following compromise:

- Revert `readDirStreamWithType` to the previous semantics, i.e. no fallback `stat`. Remove `DirStreamWithType` and related definitions.
- Provide functions like `statIfUnknownType :: FilePath -> Maybe (FilePath, DirType) -> IO (Maybe (FilePath, DirType))` that can be used together with `readDirStreamWithType`:
  ```haskell
  dir_p <- openDirStream dir
  readDirStreamWithType dir_p >>= statIfUnknownType dir
  ```

@hasufell @vdukhovni Would that be a way to go?

File: System/Posix/Directory/Common.hsc [hunk 2]
```diff
@@ -36,11 +57,149 @@ import System.Posix.Types
...
+getRealDirType _ SocketType = return SocketType
+getRealDirType _ WhiteoutType = return WhiteoutType
+getRealDirType getFileStatus _ = do
+    stat <- getFileStatus
+    return $ if | isBlockDevice stat -> BlockDeviceType
+                | isCharacterDevice stat -> CharacterDeviceType
+                | isNamedPipe stat -> NamedPipeType
+                | isRegularFile stat -> RegularFileType
+                | isDirectory stat -> DirectoryType
+                | isSymbolicLink stat -> SymbolicLinkType
```
[the remark this answers is not in the document]
Comment (mmhat) [reply System/Posix/Directory/Common.hsc]: Done.

File: System/Posix/Directory/Common.hsc [hunk 3]
```diff
@@ -73,6 +232,59 @@ foreign import ccall unsafe "HsUnix.h close"
...
+--   invocation of the callback and it will be freed automatically after. Do not
+--   pass it to the outside world!
+readDirStreamWith :: (DirEnt -> IO a) -> DirStream -> IO (Maybe a)
+readDirStreamWith f dstream = alloca
+  (\ptr_dEnt  -> readDirStreamWithPtr ptr_dEnt f dstream)
+
+-- | A version of 'readDirStreamWith' that takes a pre-allocated pointer in
+--   addition to the other arguments. This pointer is used to store the pointer
+--   to the next directory entry, if there is any. This function is intended for
+--   usecases where you need to read a lot of directory entries and want to
```
[the remark this answers is not in the document]
Comment (mmhat) [reply System/Posix/Directory/Common.hsc]: Done.

**Comment by vdukhovni** [comment]
> I'd also go with option 2 here, but I also see how a `stat` fallback can be valuable in many cases. How about the following compromise:
> 
> * Revert `readDirStreamWithType` to the previous semantics, i.e. no fallback `stat`. Remove `DirStreamWithType` and related definitions.
> * Provide functions like `statIfUnknownType :: FilePath -> Maybe (FilePath, DirType) -> IO (Maybe (FilePath, DirType))` that can be used together with `readDirStreamWithType`:

Yes, on the revert, but I don't see much point in providing the trivial function, which may not always be a good fit anyway. A user who implements some form of efficient stat relative to an open directly hardly needs help to combine it with a function that returns directory entries.

Now if you wanted to provide an efficient directory-relative stat (`statx` where available), that'd be more useful, and could perhaps be added to the supporting C code, rather that via FFI. Possibly at that point return the full stat data, no reason to throw most of it away.

**Comment by hasufell** [comment]
@vdukhovni 

I still don't see how this (without stat or similar fallbacks) provides us with a cross-platform high-level API.

We now expose the user to the internals of `d_type`, have to explain that it may sometimes be undefined (and then we silently return `CONST_DT_UNKNOWN`) or sometimes return `DT_UNKNOWN` on some filesystems if the kernel decides to do so. Then tell the user to look up stat or statx and figure out how to cook up a correct version (I got mine wrong for the first time).

It isn't as trivial as you're suggesting it is. I'm ok with exposing very thin wrappers around C FFI. I'm also ok with providing high-level API that is cross-platform.

----

So my opinion remains:

1. expose as much low-level code as possible (`readDirStreamWithPtr` and `d_type` etc.)
2. expose a high-level API that uses stat as fallback

Users who have performance needs or other requirements can build more efficient versions via the low level building blocks and are *on their own*. We don't need to provide extensive documentation for them, because they're dealing with low-level stuff.

If we can't agree on an API now, we can at least do 1. and add those building blocks for now. That seems fairly uncontroversial.

**Comment by vdukhovni** [comment]
> @vdukhovni
> 
> I still don't see how this (without stat or similar fallbacks) provides us with a cross-platform high-level API.

There are various kinds of *high-level*, the original API I took issue with exposed internal structures and was highly non-portable.  The new API does not, and is OS-platform agnostic.  It is much safer and easier to use.

> We now expose the user to the internals of `d_type`,

We are trying to report a file type, so we do need pattern synonyms for these. There's no way to avoid these.

> have to explain that it may sometimes be undefined (and then we silently return `CONST_DT_UNKNOWN`) or sometimes return `DT_UNKNOWN` on some filesystems if the kernel decides to do so.

There is no reason for having `CONST_DT_UNKNOWN` vs. `DT_UNKNOWN`, the same name needs to be used in both cases.  If a user wants to ask whether the platform supports ever reporting the type, that can be a separate API call.

> Then tell the user to look up stat or statx and figure out how to cook up a correct version (I got mine wrong for the first time).

If the purpose of exposing file types along with names is to traverse a directory tree, we could consider providing that, on top of this API, with all the fancy statx() calls where available.  I am not opposed, but the basic `dirent` API should not simulate a cheap lookup via an expensive one.

> If we can't agree on an API now, we can at least do 1. and add those building blocks for now. That seems fairly uncontroversial.

If you want to propose a specific PR, I'll review it, it should not expose non-portable structures as output, or require peek to read member fields, ...

**Comment by hasufell** [comment]
> If you want to propose a specific PR, I'll review it, it should not expose non-portable structures as output, or require peek to read member fields, ...

Not at the moment. I was saying that we can provide the low-level building blocks. Are you agreeing with that or not?

**Comment by vdukhovni** [comment]
Returning a directory name along with a DT_UNKNOWN or actual file type is the lowest-level interface we should expose in my view. There's little reason to distinguish at that layer between the filesystem not supporting file type metadata in the directory entry and the C-API not making the data available.

Users who want robust/efficient traversal based on that data will for now need to roll their own upper layers.  At some point we could offer a suitably robust iterator (making the equivalent of "rm -r" safe against hostile actors manipulating the directory tree during the traversal). That requires some care.

**Comment by hasufell** [comment]
> Returning a directory name along with a DT_UNKNOWN or actual file type is the lowest-level interface we should expose in my view.

I'm not sure if my opinion was clear enough: I'm suggesting to expose C FFI, at most a haskell function with a callback taking the DirEnt as an argument (the initial proposal from the author). Nothing else.

I'm still very clearly -1 on providing a seemingly high-level Haskell function that is non-portable.

**Comment by hasufell** [comment]
I propose to unblock this with a vote:

1. @vdukhovni suggests to expose a function that doesn't abstract over platforms that don't have `d_type` and let the caller handle it: https://github.com/haskell/unix/pull/251#pullrequestreview-1304924902

2. @hasufell proposed to expose FFI functions and a high-level function that abstracts over platforms, with a possible performance penalty on those.

I'm +1 on suggestion 2.

@vdukhovni @Bodigrim @pcapriotti @chessai @igfoo

**Comment by hasufell** [comment]
statx has been added to unix. But it doesn't work on macOS afaiu.

**Comment by vdukhovni** [comment]
> statx has been added to unix. But it doesn't work on macOS afaiu.

Not surprising, MacOS unix is a *BSD descendent. There is no statx(2) in FreeBSD or NetBSD either, and I expect ditto for OpenBSD.

**Comment by vdukhovni** [comment]
> I propose to unblock this with a vote:
> 
> 1. @vdukhovni suggests to expose a function that doesn't abstract over platforms that don't have `d_type` and let the caller handle it: [Expose dirent pointer and added readDirStreamWith #251 (review)](https://github.com/haskell/unix/pull/251#pullrequestreview-1304924902)
> 2. @hasufell proposed to expose FFI functions and a high-level function that abstracts over platforms, with a possible performance penalty on those.
> 
> I'm +1 on suggestion 2.
> 
> @vdukhovni @Bodigrim @pcapriotti @chessai @igfoo

My vote is still 1. Not fond of exposing FFI to low-level storable structures with platform-dependent fields.
The statx(2) implementation is the sort of thing that I do support.  A high-level API that's supported on a subset of platforms.
In the case of `statx(2)`, did we end up exposing a compile-time constant that makes it easy for users of `unix` to determine whether their platform has `statx(2)` without resorting to `CPP` or `try` blocks? Perhapst that should be added if not present.

**Comment by vdukhovni** [comment]
> In the case of `statx(2)`, did we end up exposing a compile-time constant that makes it easy for users of `unix` to determine whether their platform has `statx(2)` without resorting to `CPP` or `try` blocks? Perhapst that should be added if not present.

A user could approximate this by checking whether `StatxType == 0` or some other expected to be non-zero with `statx(2)` pattern synonym, but we should perhaps provide a more natural indication of support.

**Comment by hasufell** [comment]
@vdukhovni 

```hs
#include "HsUnix.h"

module Main where

main :: IO ()
main = do
#ifdef HAVE_STATX
  putStrLn "STATX"
#else
  putStrLn "Hello, Haskell!"
#endif
```

This seems to work.

**Comment by vdukhovni** [comment]
> @vdukhovni
> 
> ```haskell
> ...
> #ifdef HAVE_STATX
>   putStrLn "STATX"
> #else
>   putStrLn "Hello, Haskell!"
> #endif
> ```
> 
> This seems to work.

Sure, but I was think that it be `friendly` to our users to do that for them, and expose a boolean value they can test without any of `CPP` guards or `#include` directives. They can do that by carefully choosing which pattern synonym to compare with 0, but is ad-hoc and non-obvious.

This is not a "must have", just something that could be a useful addition, unless you see reasons not to...

**Comment by hasufell** [comment]
I'm a bit embarrassed that this didn't move forward in time.

My suggestion is to move the new API to modules that end in `.Internal` and just merge.

Any objections?

**Comment by mmhat** [comment]
@hasufell Would work for me. But I think it would be good to keep the issue open, I still hope that we expose some of the functionality in the public API in the future...

**Comment by hasufell** [comment]
Alright. If you adjust the PR to expose.via `Internal` modules, I will merge in 2 weeks, unless there are objections.

To be clear: internal modules are still under PVP, but I see no obligation to do e.g. deprecation periods for them. They may break at random.

**Comment by mmhat** [comment]
@hasufell The stuff from the `System.Posix.Directory.Common` module is now exposed in `System.Posix.Directory.Internals`. The actual implementation still lives in the `Common` since it is used in the implementations of the `readDirStreamMaybe` functions. Is that ok?

**Comment by mmhat** [comment]
@hasufell @vdukhovni Is there anything left to do for me or can we finally merge this?

**Comment by hasufell** [comment]
This has been released: https://hackage.haskell.org/package/unix-2.8.6.0

File: cbits/HsUnix.c [hunk 4]
```diff
@@ -104,6 +104,15 @@ char *__hscore_d_name( struct dirent* d )
   return (d->d_name);
 }
 
+char __hscore_d_type( struct dirent* d )
+{
+#ifdef HAVE_DIRENT_D_TYPE
```
Comment (nh2) [inline cbits/HsUnix.c:109]: Was this tested / did it work for you?

Apparently it is always false because the macro name is wrong:

https://github.com/haskell/unix/issues/347
